File: reorderLL.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
/*
 * 1) divide the linkedlist into 2 parts using lsow and fast pointer
 * 2) Reverse the 2nd half
 * 3) merge the 2 linkedlist to reorder
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null)
            return;
        // finding middle

        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // reverse the 2nd half
        ListNode current = slow.next;
        ListNode first = current.next;
        ListNode prev = null;
        slow.next = null;

        while (first != null) {
            current.next = prev;
            prev = current;
            current = first;
            first = first.next;
        }

        current.next = prev;

        // merge the 2 lists

        slow = head;
        fast = current;

        while (fast != null) {
            ListNode temp = slow.next;
            slow.next = fast;
            fast = fast.next;
            slow.next.next = temp;
            slow = temp;
        }

    }
}

File: ReorderLL.py
# // Time Complexity : O(n)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
# 1. find mid
# 2. reverse from mid.next - 
#                   1. using slow.next
#                   2. slow.next = None to detach
# 3. Merge two - Use chaining

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head == None or head.next == None: return None
        slow = head
        fast = head
        #reverse function
        def rev(cur):
            prev = None
            fast = cur.next
            while fast:
                cur.next = prev
                prev = cur
                cur = fast 
                fast = fast.next
            cur.next = prev
            return cur
        
        # find mid
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        
        second = rev(slow.next)
        # detach
        slow.next = None
        cur = ListNode(-1)
        
        #merge
        while second and head:
            #taking from first LL
            cur.next = head
            cur = cur.next
            head = head.next
            
            #taking from second LL
            cur.next = second
            cur = cur.next
            second = second.next
            
        #remaining
        while second:
            cur.next = second
            cur = cur.next
            second = second.next
            
        #remaining
        while head:
            cur.next = head
            cur = cur.next
            head = head.next
        

            
            
        
        